### PR TITLE
CHIA-194: CHIP-0026 Mempool Updates

### PR DESCRIPTION
**Pull request by Rigidity**
### Purpose:

Implements mempool updates described in [CHIP-0026](https://github.com/Chia-Network/chips/pull/100).

### Current Behavior:

There is no way to receive updates on specifically subscribed transactions entering and leaving the mempool.

### New Behavior:

Now you can opt in to mempool updates for subscriptions, which makes transaction lists and replace by fee easier in wallets.

### Testing Notes:

Not yet tested.

Manually tested on Testet11 and Mainnet, and an extensive suite of unit tests covering all of the different mempool update features.

### CODE REVIEW

File: chia/full_node/mempool.py [hunk 1]
```diff
@@ -155,6 +168,58 @@ def all_item_ids(self) -> List[bytes32]:
...
+                found_addition = False
+
+                for puzzle_hash, amount, _memo in spend.create_coin:
+                    if Coin(spend.coin_id, puzzle_hash, uint64(amount)).name() in coin_ids:
+                        transaction_ids.append(transaction_id)
+                        found_addition = True
+                        break
+
+                if found_addition:
+                    break
```
Comment (arvidn) [inline chia/full_node/mempool.py:193]:
```suggestion
                else:
                    continue
                break
```
Instead of using `found_additions` as state to know whether to exit the next level loop, you can use `for ... else`

File: chia/full_node/mempool.py [hunk 2]
```diff
@@ -155,6 +168,58 @@ def all_item_ids(self) -> List[bytes32]:
...
+                found_addition = False
+
+                for puzzle_hash, _amount, memo in spend.create_coin:
+                    if puzzle_hash in puzzle_hashes or (include_hints and memo is not None and memo in puzzle_hashes):
+                        transaction_ids.append(transaction_id)
+                        found_addition = True
+                        break
+
+                if found_addition:
+                    break
```
Comment (arvidn) [inline chia/full_node/mempool.py:219]: same thing here. you could drop the `found_addition` variable by using `for ... else` instead

File: chia/full_node/mempool.py [hunk 3]
```diff
@@ -155,6 +168,58 @@ def all_item_ids(self) -> List[bytes32]:
...
+
+        return transaction_ids
+
+    def transaction_ids_matching_puzzle_hashes(self, puzzle_hashes: Set[bytes32], include_hints: bool) -> List[bytes32]:
```
Comment (arvidn) [inline chia/full_node/mempool.py:197]: I think at least this function and `transaction_ids_matching_coin_ids()` should have unit tests in `test_mempool.py`. Ideally the other changes in `Mempool` should have tests too, like, the information you're returning

File: chia/full_node/mempool_manager.py [hunk 4]
```diff
@@ -748,22 +773,23 @@ async def local_get_coin_records(names: Collection[bytes32]) -> List[CoinRecord]
...
+                txs_added.append(NewPeakItem(item.spend_bundle_name, item.spend_bundle, item.npc_result))
+            mempool_item_removals.extend(info.removals)
         log.info(
             f"Size of mempool: {self.mempool.size()} spends, "
             f"cost: {self.mempool.total_mempool_cost()} "
             f"minimum fee rate (in FPC) to get in for 5M cost tx: {self.mempool.get_min_fee_rate(5000000)}"
         )
         self.mempool.fee_estimator.new_block(FeeBlockInfo(new_peak.height, included_items))
-        return txs_added
+        return NewPeakInfo(txs_added, mempool_item_removals)
```
Comment (arvidn) [inline chia/full_node/mempool_manager.py:792]: I think you should have unit tests to ensure you get the correct removals returned here and in the other plaes if `MempoolManager`

File: chia/full_node/subscriptions.py [hunk 5]
```diff
@@ -208,3 +212,31 @@ def coin_subscription_count(self) -> int:
 
     def puzzle_subscription_count(self) -> int:
         return self._puzzle_subscriptions.total_count()
+
+
+def peers_for_spend_bundle(
+    peer_subscriptions: PeerSubscriptions, conds: SpendBundleConditions, hints_for_removals: Set[bytes32]
```
Comment (arvidn) [inline chia/full_node/subscriptions.py:218]: I think this function should have a unit test. A doc-comment would be nice as well

File: chia/_tests/wallet/test_new_wallet_protocol.py [hunk 6]
```diff
@@ -20,21 +24,34 @@
...
     full_node_api = full_node_service._api
     fn_server = full_node_api.server
 
-    incoming_queue, peer_id = await add_dummy_connection(fn_server, self_hostname, 41723, NodeType.WALLET)
+    incoming_queue, peer_id = await add_dummy_connection(
+        fn_server,
+        self_hostname,
+        41723,
+        NodeType.WALLET,
+        additional_capabilities=[(uint16(Capability.MEMPOOL_UPDATES), "1")] if mempool_updates else [],
```
[the remark this answers is not in the document]
Comment (Rigidity) [reply chia/_tests/wallet/test_new_wallet_protocol.py:53]: Not on the wallet, but I can make it always enabled in this file without issue

**Review by arvidn — APPROVED** [review]
nice!

**Comment by Rigidity** [comment]
The missing coverage is due to:
* Edge cases which would be very difficult to test for (`peer is None` due to a race condition)
* Code paths which were already uncovered but modified slightly

The mempool updates feature is fully tested
